bindings/python/verovio.i

```diff
@@ -27,142 +29,140 @@
 // Toolkit::Edit
 %feature("shadow") vrv::Toolkit::Edit( const std::string & ) %{
 def edit(toolkit, editor_action):
+    """Edit the MEI data."""
```

Comment:
Do these need to be hand-edited? If the doc for the C++ is updated, will it be updated here too, or are there two places?

Reply:
Unfortunately I don't see another way as copying them here manually.

Reply:
Turns out swig is able to use doxygen for generating docstring (although only for functions that are not overloaded): https://www.swig.org/Doc4.0/Doxygen.html#Doxygen

Replaced the deprecated `-py3` option with `-doxygen` option in 937bb8f3b551658e1b72a325db5e269bd9c5ebba

Reply:
```console
>>> print(tk.renderToSVG.__doc__)

        Render a page to SVG.

        :type pageNo: int, optional
        :param pageNo: The page to render (1-based)
        :type xmlDeclaration: boolean, optional
        :param xmlDeclaration: True for including the xml declaration in the SVG output
        :rtype: string
        :return: The SVG page as a string
        
```

Reply:
Not sure what to do with the duplicated documentation of the overloaded methods. Maybe we can just point to the reference book? I made sure the ones there are all coming from the source code.

Reply:
From what I understand the docstrings in the SWIG file can be removed, and SWIG will automatically add them to the Python methods. Or maybe I'm wrong?

Reply:
Does this also need to happen here: 

https://github.com/rism-digital/verovio/blob/develop/setup.py#L121

Reply:
Well, I tried removing them from the swig file, but that will just create empty docstrings. Maybe we'll come up with a better solution later.

Reply:
So... 

It seems that the docstrings are missing from the shadowed Python methods because those override it. However, this is fixable. 

Rather than, for example:

```
%feature("shadow") vrv::Toolkit::GetAvailableOptions() const %{
def getAvailableOptions(toolkit) -> dict:
     return json.loads($action(toolkit))
%}
```

This can be changed to:

```
%feature("pythonappend") vrv::Toolkit::GetAvailableOptions() const %{
    val = json.loads(val)
%}
```

It seems that the auto-generated Python proxy method sets the value of `$action` to the `val` variable, so you can intercept that before the return and modify it.

This generates a method that looks like this:

```
    def getAvailableOptions(self):
        r"""
        Return all available options grouped by category

        For each option, returns the type, the default value, and the minimum and maximum value (when available)

        :rtype: string
        :return: A stringified JSON object
        """
        val = _verovio_module_3_11_0.toolkit_getAvailableOptions(self)

        val = json.loads(val)


        return val
```

I'm not super happy about the value of `val` being reset, but _c'est la vie_ I guess.

Reply:
For the shadowed methods that have stuff happening before the C++ call, there is also `%pythonprepend`, which does the same except the arguments are held in an `args[]` list.

Reply:
Slight correction -- not held in the `args[]` list but set to a variable named after the C++ variable name. So:

```
%feature("pythonprepend") vrv::Toolkit::SetOptions(const std::string &) %{
    jsonOptions = json.dumps(jsonOptions)
%}
```

generates:

```
    def setOptions(self, jsonOptions):
        r"""
        Set option values

        The name of each option to be set is to be given as JSON key.

        :type jsonOptions: string
        :param jsonOptions: A stringified JSON objects with the output options
        :rtype: boolean
        :return: True if the options were successfully set
        """

        jsonOptions = json.dumps(jsonOptions)


        return _verovio_module_3_11_0.toolkit_setOptions(self, jsonOptions)
```

Reply:
I guess I went of the %feature("shadow") because in some methods we need to do adjustments before **and** after.

Reply:
Yes, I remember reading that.

If we can do both then we should give a try. @rettinghaus are you willing to try this?

Reply:
I'm on it. 

Any specific reason `GetDescriptiveFeatures` takes `options` instead of `jsonOptions`?

Reply:
No, It can be renamed.